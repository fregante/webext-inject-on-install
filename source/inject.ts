import {injectContentScript, isScriptableUrl} from 'webext-content-scripts';
import {isPersistentBackgroundPage} from 'webext-detect';
import chromeP from 'webext-polyfill-kinda';

const acceptableInjectionsCount = 10;

type ContentScript = NonNullable<chrome.runtime.Manifest['content_scripts']>[number];

export const tracked = new Map<number, ContentScript[]>();

function forgetTab(tabId: number) {
	tracked.delete(tabId);
	if (tracked.size === 0) {
		chrome.tabs.onUpdated.removeListener(onDiscarded);
		chrome.tabs.onRemoved.removeListener(forgetTab);
		chrome.tabs.onActivated.removeListener(onActivated);
		chrome.webNavigation?.onCommitted.removeListener(onCommitted);
	}
}

function onDiscarded(tabId: number, changeInfo: {discarded?: boolean}) {
	if (changeInfo.discarded) {
		forgetTab(tabId);
	}
}

function onCommitted({tabId, frameId}: {tabId: number; frameId: number}) {
	if (frameId === 0) {
		forgetTab(tabId);
	}
}

function onActivated({tabId}: {tabId: number}) {
	const scripts = tracked.get(tabId);
	if (!scripts) {
		return;
	}

	forgetTab(tabId);
	console.debug('webext-inject-on-install: Deferred injection', scripts, 'into tab', tabId);
	for (const script of scripts) {
		void injectContentScript(tabId, script);
	}
}

export default async function progressivelyInjectScript(contentScript: ContentScript) {
	const permissions = globalThis.chrome?.runtime.getManifest().permissions;
	if (!permissions?.includes('tabs')) {
		throw new Error('webext-inject-on-install: The "tabs" permission is required');
	}

	const liveTabs = await chromeP.tabs.query({
		url: contentScript.matches,
		discarded: false,

		// Excludes unloaded tabs https://github.com/fregante/webext-inject-on-install/issues/11
		status: 'complete',
	});

	// `tab.url` is empty when the browser is starting, which is convenient because we don't need to inject anything.
	const scriptableTabs = liveTabs.filter(tab => isScriptableUrl(tab.url));
	console.debug('webext-inject-on-install: Found', scriptableTabs.length, 'tabs matching', contentScript);

	if (scriptableTabs.length === 0) {
		return;
	}

	// TODO: Non-persistent pages support via chrome.storage.session
	// https://github.com/fregante/webext-inject-on-install/issues/4
	const singleInjection = !isPersistentBackgroundPage() || scriptableTabs.length <= acceptableInjectionsCount;
	console.debug('webext-inject-on-install: Single injection?', singleInjection);

	for (const tab of scriptableTabs) {
		if (singleInjection || tab.active) {
			console.debug('webext-inject-on-install: Injecting', contentScript, 'into tab', tab.id);
			void injectContentScript(
				// Unless https://github.com/fregante/webext-content-scripts/issues/30 is changed
				contentScript.all_frames ? tab.id! : {tabId: tab.id!, frameId: 0},
				contentScript,
			);
		} else {
			chrome.tabs.onUpdated.addListener(onDiscarded);
			chrome.tabs.onRemoved.addListener(forgetTab);
			chrome.tabs.onActivated.addListener(onActivated);

			// Catch tab navigations that happen while the tab is not active
			chrome.webNavigation?.onCommitted.addListener(onCommitted);

			// eslint-disable-next-line @typescript-eslint/no-unnecessary-type-assertion -- Uh, wrong?
			const scripts = tracked.get(tab.id!) ?? [];
			scripts.push(contentScript);

			// eslint-disable-next-line @typescript-eslint/no-unnecessary-type-assertion -- Uh, wrong?
			tracked.set(tab.id!, scripts);
		}
	}
}
