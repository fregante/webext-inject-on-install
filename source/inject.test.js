
import {
	expect, test, beforeEach, vi,
} from 'vitest';
import {chrome} from 'vitest-chrome';
import {injectOneScript} from './inject.js';
import manifest from './demo/manifest.json' with {type: 'json'};

beforeEach(() => {
	vi.clearAllMocks();
	chrome.runtime.getManifest.mockReturnValue(manifest);
	chrome.storage.session.get.mockResolvedValue({});
});

test('base usage', async () => {
	const contentScript = manifest.content_scripts[0];
	const scriptableTabs = [
		{
			url: 'https://example.com/', discarded: false, id: 1, active: true,
		},
	];

	chrome.tabs.query.mockImplementation(async () => scriptableTabs);

	await injectOneScript(contentScript);
	expect(chrome.tabs.executeScript.mock.lastCall[1]).toMatchSnapshot();
	expect(chrome.tabs.insertCSS.mock.calls.length).toBe(0);
	expect(chrome.tabs.onUpdated.hasListeners()).toBe(false);
	expect(chrome.tabs.onUpdated.hasListeners()).toBe(false);
	expect(chrome.tabs.onRemoved.hasListeners()).toBe(false);
	expect(chrome.tabs.onActivated.hasListeners()).toBe(false);
});

test('css content', async () => {
	const contentScript = manifest.content_scripts[1];
	const scriptableTabs = [
		{
			id: 1,
			url: 'https://ephiframe.vercel.app/',
			discarded: false,
			active: true,
		},
	];

	chrome.tabs.query.mockImplementation(async () => scriptableTabs);

	await injectOneScript(contentScript);
	expect(chrome.tabs.executeScript.mock.calls.length).toBe(0);
	expect(chrome.tabs.insertCSS.mock.lastCall[1]).toMatchSnapshot();
});

test('deferred usage', async () => {
	const contentScript = manifest.content_scripts[0];
	const scriptableTabs = Array.from({length: 11}).fill(0).map((_, id) => ({
		id,
		url: `https://ephiframe.vercel.app/${id}`,
		discarded: false,
		active: false,
	}));

	chrome.tabs.query.mockImplementation(async () => scriptableTabs);

	await injectOneScript(contentScript);

	// Ensure no injections were made because no matching tabs were found
	expect(chrome.tabs.executeScript.mock.calls.length).toBe(0);
	expect(chrome.tabs.insertCSS.mock.calls.length).toBe(0);
	expect(chrome.tabs.onUpdated.hasListeners()).toBe(true);
	expect(chrome.tabs.onRemoved.hasListeners()).toBe(true);
	expect(chrome.tabs.onActivated.hasListeners()).toBe(true);

	// Ensure that the tracked list of tabs is updated when they're closed
	expect(chrome.storage.session.remove.mock.calls.length).toBe(0);
	chrome.tabs.onRemoved.callListeners(1);
	expect(chrome.storage.session.remove.mock.calls.length).toBe(1);
	chrome.tabs.onRemoved.callListeners(2);
	expect(chrome.storage.session.remove.mock.calls.length).toBe(2);
	expect(chrome.storage.session.remove.mock.calls).toMatchSnapshot();

	// Ensure that the tracked list of tabs is updated when they're discarded
	chrome.tabs.onUpdated.callListeners(3, {discarded: true});
	expect(chrome.storage.session.remove.mock.calls.length).toBe(3);
	chrome.tabs.onUpdated.callListeners(4, {discarded: true});
	expect(chrome.storage.session.remove.mock.calls.length).toBe(4);

	// Ensure that the tracked list of tabs is updated when they're activated and that they're injected
	chrome.storage.session.get.mockResolvedValue({'webext-inject-on-install:5': true});
	chrome.tabs.get.mockResolvedValue({url: 'https://ephiframe.vercel.app/5', id: 5, active: true});
	chrome.tabs.onActivated.callListeners({tabId: 5});

	// TODO: Mocking library is incomplete, not worth continuing
	// await Promise.resolve();
	// expect(chrome.storage.session.remove.mock.calls.length).toBe(5);
	// expect(chrome.tabs.executeScript.mock.calls.length).toBe(2);
	// expect(chrome.tabs.insertCSS.mock.calls.length).toBe(1);
	// expect(chrome.tabs.executeScript.mock.calls).toMatchSnapshot();
	// expect(chrome.tabs.insertCSS.mock.calls).toMatchSnapshot();

	// // Ensure that navigation away removes the tab without injecting the script
	// chrome.webNavigation.onCommitted.callListeners({tabId: 6, frameId: 2});
	// expect(chrome.storage.session.remove.mock.calls.length).toBe(6);
	// chrome.webNavigation.onCommitted.callListeners({tabId: 6, frameId: 0});
	// expect(chrome.storage.session.remove.mock.calls.length).toBe(5);
	// expect(chrome.tabs.executeScript.mock.calls.length).toBe(2);

	// // Ensure that the listeners are removed once the list is empty
	// expect(chrome.tabs.onUpdated.removeListener.mock.calls.length).toBe(0);
	// chrome.tabs.onRemoved.callListeners(7);
	// chrome.tabs.onRemoved.callListeners(8);
	// chrome.tabs.onRemoved.callListeners(9);
	// chrome.tabs.onRemoved.callListeners(10);
	// chrome.tabs.onRemoved.callListeners(11);
	// expect(chrome.tabs.onUpdated.removeListener.mock.calls.length).toBe(0);
	// chrome.tabs.onRemoved.callListeners(0);

	// expect(chrome.tabs.onUpdated.removeListener.mock.calls.length).toBe(1);
});
