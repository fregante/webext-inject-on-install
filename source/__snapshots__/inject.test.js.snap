// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`base usage 1`] = `
[
  [
    1,
    {
      "allFrames": false,
      "file": "foo.js",
      "frameId": 0,
      "matchAboutBlank": undefined,
      "runAt": undefined,
    },
    [Function],
  ],
  [
    1,
    {
      "allFrames": false,
      "file": "bar.js",
      "frameId": 0,
      "matchAboutBlank": undefined,
      "runAt": undefined,
    },
    [Function],
  ],
]
`;

exports[`base usage 2`] = `
[
  [
    1,
    {
      "allFrames": false,
      "file": "foo.css",
      "frameId": 0,
      "matchAboutBlank": undefined,
      "runAt": "document_start",
    },
    [Function],
  ],
]
`;

exports[`deferred usage 1`] = `
Map {
  0 => [
    {
      "css": [
        "foo.css",
      ],
      "js": [
        "foo.js",
        "bar.js",
      ],
      "matches": [
        "https://example.com/*",
      ],
    },
  ],
  1 => [
    {
      "css": [
        "foo.css",
      ],
      "js": [
        "foo.js",
        "bar.js",
      ],
      "matches": [
        "https://example.com/*",
      ],
    },
  ],
  2 => [
    {
      "css": [
        "foo.css",
      ],
      "js": [
        "foo.js",
        "bar.js",
      ],
      "matches": [
        "https://example.com/*",
      ],
    },
  ],
  3 => [
    {
      "css": [
        "foo.css",
      ],
      "js": [
        "foo.js",
        "bar.js",
      ],
      "matches": [
        "https://example.com/*",
      ],
    },
  ],
  4 => [
    {
      "css": [
        "foo.css",
      ],
      "js": [
        "foo.js",
        "bar.js",
      ],
      "matches": [
        "https://example.com/*",
      ],
    },
  ],
  5 => [
    {
      "css": [
        "foo.css",
      ],
      "js": [
        "foo.js",
        "bar.js",
      ],
      "matches": [
        "https://example.com/*",
      ],
    },
  ],
  6 => [
    {
      "css": [
        "foo.css",
      ],
      "js": [
        "foo.js",
        "bar.js",
      ],
      "matches": [
        "https://example.com/*",
      ],
    },
  ],
  7 => [
    {
      "css": [
        "foo.css",
      ],
      "js": [
        "foo.js",
        "bar.js",
      ],
      "matches": [
        "https://example.com/*",
      ],
    },
  ],
  8 => [
    {
      "css": [
        "foo.css",
      ],
      "js": [
        "foo.js",
        "bar.js",
      ],
      "matches": [
        "https://example.com/*",
      ],
    },
  ],
  9 => [
    {
      "css": [
        "foo.css",
      ],
      "js": [
        "foo.js",
        "bar.js",
      ],
      "matches": [
        "https://example.com/*",
      ],
    },
  ],
  10 => [
    {
      "css": [
        "foo.css",
      ],
      "js": [
        "foo.js",
        "bar.js",
      ],
      "matches": [
        "https://example.com/*",
      ],
    },
  ],
}
`;

exports[`deferred usage 2`] = `
[
  [
    5,
    {
      "allFrames": true,
      "file": "foo.js",
      "frameId": undefined,
      "matchAboutBlank": undefined,
      "runAt": undefined,
    },
    [Function],
  ],
  [
    5,
    {
      "allFrames": true,
      "file": "bar.js",
      "frameId": undefined,
      "matchAboutBlank": undefined,
      "runAt": undefined,
    },
    [Function],
  ],
]
`;

exports[`deferred usage 3`] = `
[
  [
    5,
    {
      "allFrames": true,
      "file": "foo.css",
      "frameId": undefined,
      "matchAboutBlank": undefined,
      "runAt": "document_start",
    },
    [Function],
  ],
]
`;
